accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,450 @@
+  would be a separate analyzer.
+* Shipping the platform-specific analyzer and/or annotations out-of-band.
+
+## Design
+
+### Attribute
+
+OS bindings and platforms-specific BCL APIs will be annotated using a set of
+custom attributes:
+
+```C#
+namespace System.Runtime.InteropServices
+{
+    public partial struct OSPlatform
+    {
+        // Existing properties
+        // public static OSPlatform FreeBSD { get; }
+        // public static OSPlatform Linux { get; }
+        // public static OSPlatform OSX { get; }
+        // public static OSPlatform Windows { get; }
+        public static OSPlatform Android { get; }
+        public static OSPlatform iOS { get; }
+    }
+
+    public partial static class RuntimeInformation
+    {
+        // Existing API
+        // public static bool IsOSPlatform(OSPlatform osPlatform);
+
+        // Check for the OS with a >= version comparison
+        // Used to guard APIs that were added in the given OS release.
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major);
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor);
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor, int build);
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor, int build, int revision);
+
+        // Allows checking for the OS with a < version comparison
+        // Used to guard APIs that were obsoleted or removed in the given OS release.
+        public static bool IsOSPlatformAndEarlier(OSPlatform osPlatform, int major);
+        public static bool IsOSPlatformAndEarlier(OSPlatform osPlatform, int major, int minor);
+        public static bool IsOSPlatformAndEarlier(OSPlatform osPlatform, int major, int minor, int build);
+        public static bool IsOSPlatformAndEarlier(OSPlatform osPlatform, int major, int minor, int build, int revision);
+    }
+}
+
+namespace System.Runtime.Versioning
+{
+    public abstract class OSPlatformAttribute : Attribute
+    {
+        protected OSPlatformAttribute(string osPlatform,
+                                      int major,
+                                      int minor,
+                                      int build,
+                                      int revision);
+        public string PlatformIdentifier { get; }
+        public int Major { get; }
+        public int Minor { get; }
+        public int Build { get; }
+        public int Revision { get; }
+    }
+
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=false, Inherited=true)]
```

Comment:
Wasn't it the idea to specifically allow multiple attributes?

```suggestion
                    AllowMultiple=true, Inherited=true)]
```

So that a library could do something like this:

```csharp
[AddedInOSPlatform (OSPlatform.iOS, 13)]
[AddedInOSPlatform (OSPlatform.macOS, 10, 15)]
```

and provide an API that can be used in either macOS 10.15 or iOS 13.

Reply:
Originally yes, but I scoped that out. Is that critical?

Reply:
Not sure if it's critical, but it seems pretty important to me. Most API Apple releases nowadays is added to more than one platform at the same time (at least iOS and macOS, and often tvOS and watchOS as well).

Reply:
I consider it critical to replace our current availability attributes, e.g.

https://github.com/xamarin/xamarin-macios/blob/9f28bce6b87ca6ea4cc51a0dff3e2db0064bf6ca/src/CoreFoundation/OSLog.cs#L29

we re-compile this for each platform, but we don't have to write the following on every type / method
```csharp
#if __IOS__
[AddedInOSPlatformVersion (.iOS, ...)]
#elif __TVOS__
[AddedInOSPlatformVersion (.tvOS, ...)]
#elif __WATCHOS__
[AddedInOSPlatformVersion (.watchOS, ...)]
#else 
[AddedInOSPlatformVersion (.macOS, ...)]
#endif
```

Reply:
You can see a lot of examples in how we use attributes like this in the wild here:

https://github.com/xamarin/xamarin-macios/blob/master/src/foundation.cs

Reply:
At the end of the document, we have a statement of "For example, there are some APIs that work on Windows and Linux, but not on macOS." Even for that abstract scenario, we'd need multiple attributes; right?